wgengine/magicsock/magicsock_linux.go

```diff
@@ -0,0 +1,316 @@
+		// packet immediately following; the offset is immediately
+		// following both headers.
+		bpf.LoadConstant{Dst: bpf.RegX, Val: ethHeaderSize + ipv6.HeaderLen + ipv6FragmentHeaderSize},
+
+		// If we get here, we have a UDP packet; compare with our magic
+		// number. Start by loading and comparing the first 4 bytes of
+		// the UDP packet.
+		bpf.LoadIndirect{Off: udpHeaderSize, Size: 4},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: magic1, SkipTrue: 0, SkipFalse: 3},
+
+		// Compare the next 2 bytes
+		bpf.LoadIndirect{Off: udpHeaderSize + 4, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(magic2), SkipTrue: 0, SkipFalse: 1},
+
+		// Accept the whole packet
+		bpf.RetConstant{Val: 0xFFFFFFFF},
+
+		// Skip the packet
+		bpf.RetConstant{Val: 0x0},
+	}
+)
+
+// listenDisco starts listening for disco packets using an AF_PACKET socket + a
+// BPF filter. This allows us to receive disco packets even without opening the
+// firewall; see issue 3824 for more detail.
+func (c *Conn) listenDisco() {
+	if debugDisableAF_PACKET {
+		return
+	}
+	c.listenDiscoProto(&c.bpfConn4, "IPv4", magicsockFilterV4)
+	c.listenDiscoProto(&c.bpfConn6, "IPv6", magicsockFilterV6)
+}
+
+func (c *Conn) listenDiscoProto(fdptr *int, name string, filter []bpf.Instruction) {
+	// We create our socket manually here, rather than using
+	// {ipv4,ipv6}.NewPacketConn, because they don't support SOCK_RAW.
+	fd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_RAW, int(htons(unix.ETH_P_ALL)))
+	if err != nil {
+		c.logf("magicsock: error creating raw socket for %s: %v", name, err)
+		return
+	}
+
+	*fdptr = fd
+	go func() {
+		err := c.listenPacketsWithFilter(fd, filter, c.handleDiscoPacket)
+		if err != nil {
+			c.logf("magicsock: error listening for %s disco packets with AF_PACKET: %v", name, err)
+		}
+	}()
+}
+
+func (c *Conn) closeListenDisco() {
+	if c.bpfConn4 != 0 {
+		unix.Close(c.bpfConn4)
+	}
+	if c.bpfConn6 != 0 {
+		unix.Close(c.bpfConn6)
+	}
+}
+
+// handleDiscoPacket is called by listenPacketsWithFilter when a packet
+// matching a BPF filter is received. It can be passed either an IPv4 or IPv6
+// packet and will handle either case.
+func (c *Conn) handleDiscoPacket(packet []byte, from unix.Sockaddr) error {
+	pkt := gopacket.NewPacket(packet, layers.LayerTypeEthernet, gopacket.NoCopy)
+
+	var (
+		srcIP net.IP
+		is4   bool
+	)
+	if l := pkt.Layer(layers.LayerTypeIPv6); l != nil {
+		srcIP = l.(*layers.IPv6).SrcIP
+		is4 = false
+	} else if l := pkt.Layer(layers.LayerTypeIPv4); l != nil {
+		srcIP = l.(*layers.IPv4).SrcIP
+		is4 = true
+	} else {
+		// unknown type; non-fatal
+		return nil
+	}
+
+	udpL := pkt.Layer(layers.LayerTypeUDP)
+	if udpL == nil {
+		// unexpected, but non-fatal
+		return nil
+	}
+	udp := udpL.(*layers.UDP)
+	srcPort := udp.SrcPort
+
+	srcAddr, ok := netip.AddrFromSlice(srcIP)
+	if !ok {
+		return nil
+	}
+
+	src := netip.AddrPortFrom(srcAddr, uint16(srcPort))
+	success := c.handleDiscoMessage(udp.Payload, src, key.NodePublic{})
```

Comment:
are you just handling disco messages twice via both this new path and the old path in some cases?

I figured we'd do some initial runtime probe on start-up to see if this BPF stuff was working and if so then disable the handleDiscoMessage-via-UDP path in magicsock